Update Makefile documentation

zachmargolis opened the pull request:
Based on https://github.com/18F/identity-idp/pull/5791#discussion_r779826359

zachmargolis commented:
Updated to expand out variables so we can get clearer targets:

```
--- Help ---
brakeman                            Runs brakeman
check                               Runs lint tests and spec tests
check_asset_strings                 Checks for strings
config/application.yml              Runs setup scripts (updates packages, dependencies, databases, etc)
docker_setup                        Setup script for Docker development
fast_setup                          Abbreviated setup script that skips linking some files
fast_test                           Abbreviated test run, runs RSpec tests without accessibility specs
help                                Show this help
lint                                Runs all lint tests
lint_country_dialing_codes          Checks that countries supported by Pinpoint for voice and SMS are up to date
lint_erb                            Lints ERB files
lint_optimized_assets               Checks that assets are optimized
lint_yaml                           Lints YAML files
lintfix                             Runs rubocop fix
normalize_yaml                      Normalizes YAML files (alphabetizes keys, fixes line length, smart quotes)
optimize_assets                     Optimizes all assets
optimize_svg                        Optimizes SVG images
run                                 Runs the development server
run-https                           Runs the develpment server with HTTPS
test                                Runs RSpec and yarn tests
tmp/localhost-3000.crt              Self-signed cert for local HTTPS development
update_pinpoint_supported_countries Updates list of countries supported by Pinpoint for voice and SMS
```

example: `tmp/localhost-3000.crt` is `tmp/localhost-3000.crt` in the source code

aduth commented on scripts/makefile_help.rb:
```diff
@@ -0,0 +1,40 @@
+expanded_makefile = `make --dry-run --print-data-base --no-builtin-rules`
+
+# Map of lineno => rule name run by printing the Makefile database
+expanded_targets = expanded_makefile.scan(/\n\n(?<target>[^:]+):( (?<sources>[^\n]+))?
+((#  [^\n]+)?\n?)*?
+#  commands to execute \(from `Makefile', line (?<lineno>\d+)\):/m).
+  map { |target, *, lineno| [lineno.to_i, target] }.
+  to_h
```
I see Rubocop isn't taking nicely to this.

```
scripts/makefile_help.rb:7:43: C: [Correctable] Layout/FirstMethodArgumentLineBreak: Add a line break before the first argument of a multi-line method argument list.
expanded_targets = expanded_makefile.scan(/\n\n(?<target>[^:]+):( (?<sources>[^\n]+))? ...
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
scripts/makefile_help.rb:7:43: W: Lint/MixedRegexpCaptureTypes: Do not mix named captures and numbered captures in a Regexp literal.
expanded_targets = expanded_makefile.scan(/\n\n(?<target>[^:]+):( (?<sources>[^\n]+))? ...
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

zachmargolis replied:
45bb2ab3d8e806974b5783efe22f12186bac8005

added some ignores, because this is `/m` but not `/x` regex so whitespace matters and indenting the regex causes problems

Also mixing capture groups...these helped me remember what to grab, so I am just disabling for now

aduth commented on scripts/makefile_help.rb:
```diff
@@ -0,0 +1,55 @@
+
+target_comments.each do |target, (comment, lineno)|
+  cleaned_up_name = if target.include?('$(')
+    # "tmp/$(HOST)-$(PORT).crt" into %r|tmp/.+-.+.crt|
+    rule_regexp = Regexp.new(target.gsub(/(\$\([^)]+\))/, '.+'))
```
Capture group necessary?

```suggestion
    # "tmp/$(HOST)-$(PORT).crt" into %r|tmp/.+-.+.crt|
    rule_regexp = Regexp.new(target.gsub(/\$\([^)]+\)/, '.+'))
```

Any time I see a `.+`, I also think it might be wise to make it greedy:

```suggestion
    # "tmp/$(HOST)-$(PORT).crt" into %r|tmp/.+?-.+?.crt|
    rule_regexp = Regexp.new(target.gsub(/\$\([^)]+\)/, '.+?'))
```

aduth reviewed (approved):
I like this 👍  

A couple general thoughts:

- I may worry how durable this is with different syntaxes (e.g. whitespace, variable interpolation defaulting like `$(HOST:-localhost)`.
- One of the ideas with environment variables in the targets is they can be externally customized, like how we [document to use `HOST=0.0.0.0`](https://github.com/18F/identity-idp#testing-the-application-over-https) for binding to `0.0.0.0`. This isn't reflected in the help documentation.
- I wonder if, rather than starting from `make --print-data-base` and mapping back to the original `Makefile`, it may be easier to start with the `Makefile`, and substitute variables given from `make --print-data-base`.

zachmargolis commented:
> * One of the ideas with environment variables in the targets is they can be externally customized, like how we [document to use `HOST=0.0.0.0`](https://github.com/18F/identity-idp#testing-the-application-over-https) for binding to `0.0.0.0`. This isn't reflected in the help documentation.

My issue was that it was easy to forget what the default values were, maybe the solution is to have both the template and the default, ex

```
config/application.yml [$(config)] Sets up the app
tmp/localhost-3000.crt [tmp/$(HOST)-$(PORT)] HTTPS cert for local development
```

> * I wonder if, rather than starting from `make --print-data-base` and mapping back to the original `Makefile`, it may be easier to start with the `Makefile`, and substitute variables given from `make --print-data-base`.

Maybe I should reorder the script to make that clearer, but that's technically what's happening

aduth commented:
> Maybe I should reorder the script to make that clearer, but that's technically what's happening

Ah, I think I was thrown off by the `expanded_targets`, I was thinking using the "Variables" stanza at the top of the `make --print-data-base` output, and ignoring the targets from that output.

zachmargolis commented:
Ok, big changes in 3d57d527c77b506d08d84aec3a9ae13ba40068c6!
- I rewrote the script as a class so I could unit test it
- and also got rid of the gnarly regex (good ol `.split("\n\n")`)
- and updated the output to have the template as well as the defaults:

```
--- Help ---
brakeman                                                    Runs brakeman
check                                                       Runs lint tests and spec tests
check_asset_strings                                         Checks for strings
config/application.yml              $(CONFIG)               Runs setup scripts (updates packages, dependencies, databases, etc)
docker_setup                                                Setup script for Docker development
fast_setup                                                  Abbreviated setup script that skips linking some files
fast_test                                                   Abbreviated test run, runs RSpec tests without accessibility specs
help                                                        Show this help
lint                                                        Runs all lint tests
lint_country_dialing_codes                                  Checks that countries supported by Pinpoint for voice and SMS are up to date
lint_erb                                                    Lints ERB files
lint_optimized_assets                                       Checks that assets are optimized
lint_yaml                                                   Lints YAML files
lintfix                                                     Runs rubocop fix
normalize_yaml                                              Normalizes YAML files (alphabetizes keys, fixes line length, smart quotes)
optimize_assets                                             Optimizes all assets
optimize_svg                                                Optimizes SVG images
run                                                         Runs the development server
run-https                                                   Runs the development server with HTTPS
setup                                                       Runs setup scripts (updates packages, dependencies, databases, etc)
test                                                        Runs RSpec and yarn tests
tmp/localhost-3000.crt              tmp/$(HOST)-$(PORT).crt Self-signed cert for local HTTPS development
tmp/localhost-3000.key              tmp/$(HOST)-$(PORT).key Self-signed cert for local HTTPS development
update_pinpoint_supported_countries                         Updates list of countries supported by Pinpoint for voice and SMS
```

aduth commented on lib/makefile_help_parser.rb:
```diff
@@ -0,0 +1,115 @@
+  # Map of target => [comment, lineno]
+  # target might have variables like $(HOST)
+  # @return [Hash<String, Array(String, Integer)>]
+  def target_comments
```
Should we memoize this if it's called for every parsed target?

zachmargolis replied:
oh whoops yes! fixed in ef9a5554be301723ed57be88ea2b883ab451cbb4

aduth reviewed (approved):
Looks good 👍 

Maybe want to rebase after #5795 to make sure it's still in working shape?

zachmargolis commented:
> Maybe want to rebase after #5795 to make sure it's still in working shape?

Yup! I merged the branches locally to make sure they still worked and made some recommendations in https://github.com/18F/identity-idp/pull/5795#discussion_r780474644
